aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,884 @@
+  TORCH_CUDASPARSE_CHECK(cusparseCreateCsrgemm2Info(&gemm2Info_));
+
+  csrOutput out({A.size(0), B.size(1)});
+
+  int innerSize = confirm_mult_size(A.size_, B.size_);
+
+  out.csr_pointers_ = at::empty({out.size(0) + 1}, output_indices.options().dtype(kInt));
+
+  // Compute needed buffer size
+  size_t new_bubber_sz;
+  TORCH_CUDASPARSE_CHECK(cusparseScsrgemm2_bufferSizeExt(
+      cusparseHandle_,
+      out.size(0),
+      out.size(1),
+      innerSize,
+      alpha,
+      A.description_,
+      A.nnz_,
+      A.csr_pointers_,
+      A.csr_indices_,
+      B.description_,
+      B.nnz_,
+      B.csr_pointers_,
+      B.csr_indices_,
+      beta,
+      C.description_,
+      C.nnz_,
+      C.csr_pointers_,
+      C.csr_indices_,
+      gemm2Info_,
+      &new_bubber_sz));
+
+  // (Re)allocate buffer if needed
+  if (new_bubber_sz > currentBufferSize_) {
+    if (buffer_ != NULL) {
+      cudaFree(buffer_);
+    }
+    cudaMalloc(&buffer_, new_bubber_sz);
+    currentBufferSize_ = new_bubber_sz;
+  }
+
+  // Find the resulting non-zero pattern.
+  TORCH_CUDASPARSE_CHECK(cusparseXcsrgemm2Nnz(
+      cusparseHandle_,
+      out.size(0),
+      out.size(1),
+      innerSize,
+      A.description_,
+      A.nnz_,
+      A.csr_pointers_,
+      A.csr_indices_,
+      B.description_,
+      B.nnz_,
+      B.csr_pointers_,
+      B.csr_indices_,
+      C.description_,
+      C.nnz_,
+      C.csr_pointers_,
+      C.csr_indices_,
+      out.description_,
+      out.csr_pointers_.data_ptr<int>(),
+      &out.nnz_,
+      gemm2Info_,
+      buffer_));
+
+  out.csr_indices_ = at::empty({out.nnz_}, output_indices.options().dtype(kInt));
+  out.csr_values_ = at::empty({out.nnz_}, output_values.options());
+
+  // Perform the gemm2 operation for doubles
+  // out = alpha ∗ A ∗ B + beta ∗ C
+  TORCH_CUDASPARSE_CHECK(cusparseScsrgemm2(
+      cusparseHandle_,
+      out.size(0),
+      out.size(1),
+      innerSize,
+      alpha,
+      A.description_,
+      A.nnz_,
+      A.csr_values_,
+      A.csr_pointers_,
+      A.csr_indices_,
+      B.description_,
+      B.nnz_,
+      B.csr_values_,
+      B.csr_pointers_,
+      B.csr_indices_,
+      beta,
+      C.description_,
+      C.nnz_,
+      C.csr_values_,
+      C.csr_pointers_,
+      C.csr_indices_,
+      out.description_,
+      out.csr_values_.data_ptr<float>(),
+      out.csr_pointers_.data_ptr<int>(),
+      out.csr_indices_.data_ptr<int>(),
+      gemm2Info_,
+      buffer_));
+
+  TORCH_CUDASPARSE_CHECK(cusparseDestroy(cusparseHandle_));
+  TORCH_CUDASPARSE_CHECK(cusparseDestroyCsrgemm2Info(gemm2Info_));
+  return out;
+}
+
+
+csrOutput Dgemm2(
+    const DcsrMatrixRef& A,
+    const DcsrMatrixRef& B,
+    const DcsrMatrixRef& C,
+    const double* alpha,
+    const double* beta,
+    Tensor &output_values, 
+    IntTensor &output_indices) {
+  cusparseHandle_t cusparseHandle_;
+  csrgemm2Info_t gemm2Info_;
+  static void* buffer_{nullptr};
```

Comment:
Please don't do this. If you use the caching allocator you will bypass most of the overhead you were trying to solve here. By permanently storing the workspace in a static variable, you permanently leak the workspace memory the first time someone calls into sparse matmul, and there is no way for the user to reclaim it.